### PR TITLE
Docker builds - refactor

### LINKED ISSUE #7574: Docker builds - refactor
Proposed changes:
- Switch to Python 3.8 and Ubuntu 20.04 as a base image
- Refactor Dockerfiles in order to manage cache better
- Build a base image with MITIE
- Tag the latest stable version as `latest`
- Tag Docker images build against the master branch as `master`
- Add options for building Docker images locally to the Makefile

### PR DESCRIPTION
**Pull request by tczekajlo**
**Proposed changes**:
- Resolves #7574
- Use Ubuntu 20.04 as a base image
- Build a base image with MITIE
- Tag the latest stable version as `latest`
- Tag Docker images build against the master branch as `master`
- Add options for building Docker images locally to the Makefile

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tczekajlo** [comment]
@m-vdb @wochinge The time came for the rasa repository and refactor of Docker builds. The PR is ready to review, can you take a look at it if you have a minute?

**Review by m-vdb — COMMENTED** [review]
Amazing work @tczekajlo 🚀 🚀 Here's my review, I think I'm more and more comfortable with Docker / buildx and that's thanks to you ahah!

File: .github/workflows/continous-integration.yml [hunk 1]
```diff
@@ -279,27 +281,170 @@ jobs:
...
+        if [[ "${IS_NEWEST_VERSION}" == "true" && "$CURRENT_TAG" =~ ^[0-9.]+$ ]]; then
+          echo "::set-output name=is_newest_version::true"
+        else
+          echo "::set-output name=is_newest_version::false"
+        fi
+
+    - name: Check if a base image exists
+      id: check_image
+      run: |
+        DOCKERHUB_TAGS_URL="https://registry.hub.docker.com/v2/repositories/rasa/rasa/tags?page_size=10000"
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply .github/workflows/continous-integration.yml:340]: It's a good idea 👍🏻

File: .github/workflows/continous-integration.yml [hunk 2]
```diff
@@ -312,14 +457,16 @@ jobs:
         docker rmi $(docker image ls -aq)
         df -h
 
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v1
+      with:
+        version: latest
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply .github/workflows/continous-integration.yml:463]: Good point, I'll pin it to a specific version.

File: .github/workflows/continous-integration.yml [hunk 3]
```diff
@@ -328,22 +475,38 @@ jobs:
...
-      run: docker build . --file ${{ matrix.image.file }} --tag rasa/rasa:latest${{ matrix.image.tag_ext }} --cache-from rasa/rasa:latest${{ matrix.image.tag_ext }}
+      run: |
+        docker buildx bake -f docker/docker-bake.hcl ${{ matrix.image }}
 
-    - name: Push image with latest tag 📦
+    - name: Push image with master tag 📦
       if: needs.changes.outputs.docker == 'true' && github.event_name == 'push' && github.ref == 'refs/heads/master' && github.repository == 'RasaHQ/rasa'
-      run: docker push rasa/rasa:latest${{ matrix.image.tag_ext }}
+      run: |
+        docker buildx bake -f docker/docker-bake.hcl ${{ matrix.image }} --push
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply .github/workflows/continous-integration.yml:486]: The image will already exist, so it doesn't rebuild, only push.

File: Makefile [hunk 4]
```diff
@@ -178,3 +178,33 @@ livedocs:
 
 release:
 	poetry run python scripts/release.py
+
+build-docker:
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply Makefile:182]: Yep, the warning was removed in 0.5.0 - https://github.com/docker/buildx/pull/390

File: .github/workflows/continous-integration.yml [hunk 5]
```diff
@@ -279,27 +281,170 @@ jobs:
...
+      if: github.event_name == 'pull_request'
+      with:
+        ref: ${{ github.event.pull_request.head.sha }}
+
     - name: Checkout git repository 🕝
-      if: needs.changes.outputs.docker == 'true'
       uses: actions/checkout@v2
+      if: github.event_name != 'pull_request'
 
     - name: Free disk space
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply .github/workflows/continous-integration.yml:449]: Yes :)

File: docker/Dockerfile.base-mitie [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# The base image used for all images that require a MITIE model
+FROM alpine:latest
+
+# download mitie model
+RUN mkdir -p /build/data/ && wget -P /build/data/ https://github.com/mit-nlp/MITIE/releases/download/v0.4/MITIE-models-v0.2.tar.bz2 && \
```
[the remark this answers is not in the document]
Comment (tczekajlo) [reply docker/Dockerfile.base-mitie]: Ok, here I re-use the Makefile :)

**Comment by tczekajlo** [comment]
@m-vdb Thanks a lot for reviewing. It's ready for the second round, can you give another look if you have a minute?

File: .github/workflows/continous-integration.yml [hunk 7]
```diff
@@ -283,27 +285,159 @@ jobs:
...
+    # IMAGE_TAG=2.0.6
+    - name: Set environment variables - push - tag
+      if: github.event_name == 'push' && env.IS_TAG_BUILD == 'true'
+      run: |
+        TAG_NAME=${GITHUB_REF#refs/tags/}
+        echo "IMAGE_TAG=${TAG_NAME}" >> $GITHUB_ENV
+
+    # Set environment variables for a branch
+    #
+    # In this scenario, we've pushed changes into the master branch
```
Comment (wochinge) [inline .github/workflows/continous-integration.yml:399]:
```suggestion
    # In this scenario, we've pushed changes into the main branch
```

Comment (wochinge) [reply .github/workflows/continous-integration.yml:399]: Maxime is currently renaming that. It's not merged yet, but let's try to not introduce it again.

File: .github/workflows/continous-integration.yml [hunk 8]
```diff
@@ -283,27 +285,159 @@ jobs:
...
+    #
+    # Example output:
+    # IMAGE_TAG=master
+    - name: Set environment variables - push - branch
+      if: github.event_name == 'push' && env.IS_TAG_BUILD == 'false'
+      run: |
+        BRANCH_NAME=${GITHUB_REF#refs/heads/}
+        SAFE_BRANCH_NAME=$(echo ${GITHUB_REF#refs/heads/} | sed 's/\./-/g')
```
Comment (wochinge) [inline .github/workflows/continous-integration.yml:407]: should we use quotes around the branch names? I remember what happened when Joe had a branch name starting with `#` 😂

File: README.md [hunk 9]
```diff
@@ -202,10 +202,20 @@ Just execute this command to resolve merge conflicts in `poetry.lock` automatica
 poetry-merge-lock
 ```
 
+### Build a Docker image locally
+
+In order to build a Docker image on your local machine execute the following command:
+
+```bash
+make build-docker
```
Comment (wochinge) [inline README.md:210]: ❤️

File: docker/Dockerfile.full [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+COPY . /build/
+COPY docker/configs/config_pretrained_embeddings_spacy_en_duckling.yml /build/config.yml
+
+# change working directory
+WORKDIR /build
+
+# install dependencies
+RUN python -m venv /opt/venv && \
+  . /opt/venv/bin/activate && \
+  pip install --no-cache-dir -U 'pip<20'
```
Comment (wochinge) [inline docker/Dockerfile.full:22]: I'm pretty sure `pip<20` is no longer true and we can dump that. There was an issue in Slack that the latest latest version doesn't work, but `>20` definitely works. Also fine to do in a separate PR though.

File: docker/Dockerfile.full [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+# Create a volume for temporary data
+VOLUME /tmp
+
+# change shell
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# the entry point
+EXPOSE 5005
+ENTRYPOINT ["rasa"]
```
Comment (wochinge) [inline docker/Dockerfile.full:60]: should we use `tini`?

**Review by wochinge — COMMENTED** [review]
I'm excited. Just had a short look as Maxime has it covered 👍🏻

File: docker/Dockerfile.full (hunk 11 above)
Comment (tczekajlo) [reply docker/Dockerfile.full:60]: Here it's not needed, we don't use a shell to execute the rasa command.

File: docker/Dockerfile.full (hunk 10 above)
Comment (tczekajlo) [reply docker/Dockerfile.full]: 👍🏻

File: .github/workflows/continous-integration.yml (hunk 8 above)
Comment (tczekajlo) [reply .github/workflows/continous-integration.yml]: Good point! Thanks 🙌🏻

File: docker/Dockerfile.full (hunk 11 above)
Comment (wochinge) [reply docker/Dockerfile.full:60]: makes sense 👍🏻 Thanks for the explanation!

**Review by m-vdb — APPROVED** [review]

**Comment by m-vdb** [comment]
Fixing failing doc test in: https://github.com/RasaHQ/rasa/pull/7773
